From 9cf2d36f8181b9c0cb95c85600f4726e0cb8810a Mon Sep 17 00:00:00 2001
From: Jerome Brunet <jbrunet@baylibre.com>
Date: Mon, 22 Nov 2021 13:49:17 +0100
Subject: [PATCH 2/2] mmc: meson-gx: align clock settings on linux

Align configuration and clock registers settings with linux:
* clear CFG_CLK_AUTO so output clock is kept on when core clock is on
* set CFG_ERR_ABORT to transactions on error
* set CLK_ALWAYS_ON to keep the core clock on

The last settings depends on the ip version which is why platform
data handling is refactored here

Signed-off-by: Jerome Brunet <jbrunet@baylibre.com>
---
 drivers/mmc/meson_gx_mmc.c | 30 +++++++++++++++++++++---------
 drivers/mmc/meson_gx_mmc.h |  7 +++++++
 2 files changed, 28 insertions(+), 9 deletions(-)

diff --git a/drivers/mmc/meson_gx_mmc.c b/drivers/mmc/meson_gx_mmc.c
index 7c0827b9049e..b52dd55abdf0 100644
--- a/drivers/mmc/meson_gx_mmc.c
+++ b/drivers/mmc/meson_gx_mmc.c
@@ -17,13 +17,16 @@
 #include <linux/log2.h>
 #include "meson_gx_mmc.h"
 
-bool meson_gx_mmc_is_compatible(struct udevice *dev,
-				enum meson_gx_mmc_compatible family)
-{
-	enum meson_gx_mmc_compatible compat = dev_get_driver_data(dev);
-
-	return compat == family;
-}
+struct meson_gx_mmc_version_data meson_gx_mmc_version[] = {
+	[MMC_COMPATIBLE_V2] = {
+		.clk_always_on = BIT(24),
+		.core_phase = CLK_CO_PHASE_180,
+	},
+	[MMC_COMPATIBLE_V3] = {
+		.clk_always_on = BIT(28),
+		.core_phase = CLK_CO_PHASE_270,
+	},
+};
 
 static inline void *get_regbase(const struct mmc *mmc)
 {
@@ -44,12 +47,19 @@ static inline void meson_write(struct mmc *mmc, uint32_t val, int offset)
 
 static void meson_mmc_config_clock(struct mmc *mmc)
 {
+	struct meson_mmc_plat *pdata = mmc->priv;
 	uint32_t meson_mmc_clk = 0;
 	unsigned int clk, clk_src, clk_div;
 
 	if (!mmc->clock)
 		return;
 
+	/* Clk always on */
+	meson_mmc_clk |= pdata->version->clk_always_on;
+
+	/* Core phase according to ip version */
+	meson_mmc_clk |= pdata->version->core_phase;
+
 	/* TOFIX This should use the proper clock taken from DT */
 
 	/*
@@ -236,6 +246,7 @@ static const struct dm_mmc_ops meson_dm_mmc_ops = {
 
 static int meson_mmc_of_to_plat(struct udevice *dev)
 {
+	enum meson_gx_mmc_compatible compat = dev_get_driver_data(dev);
 	struct meson_mmc_plat *pdata = dev_get_plat(dev);
 	fdt_addr_t addr;
 
@@ -244,6 +255,7 @@ static int meson_mmc_of_to_plat(struct udevice *dev)
 		return -EINVAL;
 
 	pdata->regbase = (void *)addr;
+	pdata->version = &meson_gx_mmc_version[compat];
 
 	return 0;
 }
@@ -299,8 +311,8 @@ static int meson_mmc_probe(struct udevice *dev)
 
 	/* enable auto clock mode */
 	val = meson_read(mmc, MESON_SD_EMMC_CFG);
-	val &= ~CFG_SDCLK_ALWAYS_ON;
-	val |= CFG_AUTO_CLK;
+	val &= ~(CFG_SDCLK_ALWAYS_ON | CFG_AUTO_CLK);
+	val |= CFG_ERR_ABORT;
 	meson_write(mmc, val, MESON_SD_EMMC_CFG);
 
 	return 0;
diff --git a/drivers/mmc/meson_gx_mmc.h b/drivers/mmc/meson_gx_mmc.h
index 21d81b4734bd..08c4c0bdb936 100644
--- a/drivers/mmc/meson_gx_mmc.h
+++ b/drivers/mmc/meson_gx_mmc.h
@@ -49,6 +49,7 @@ enum meson_gx_mmc_compatible {
 #define   CFG_RC_CC_16			(4 << 12)
 #define   CFG_SDCLK_ALWAYS_ON		BIT(18)
 #define   CFG_AUTO_CLK			BIT(23)
+#define   CFG_ERR_ABORT			BIT(27)
 
 #define MESON_SD_EMMC_STATUS		0x48
 #define   STATUS_MASK			GENMASK(15, 0)
@@ -84,7 +85,13 @@ enum meson_gx_mmc_compatible {
 #define MESON_SD_EMMC_CMD_RSP2		0x64
 #define MESON_SD_EMMC_CMD_RSP3		0x68
 
+struct meson_gx_mmc_version_data {
+	uint32_t clk_always_on;
+	uint32_t core_phase;
+};
+
 struct meson_mmc_plat {
+	struct meson_gx_mmc_version_data *version;
 	struct mmc_config cfg;
 	struct mmc mmc;
 	void *regbase;
-- 
2.36.1

